src/domain_separated/constraints.rs

```diff
@@ -0,0 +1,112 @@
+use crate::constraints::AbsorbableGadget;
+use crate::constraints::CryptographicSpongeVar;
+use crate::domain_separated::{DomainSeparatedSponge, DomainSeparator};
+use crate::{Absorbable, CryptographicSponge, FieldElementSize};
+use ark_ff::PrimeField;
+use ark_nonnative_field::NonNativeFieldVar;
+use ark_r1cs_std::bits::boolean::Boolean;
+use ark_r1cs_std::bits::uint8::UInt8;
+use ark_r1cs_std::fields::fp::FpVar;
+use ark_relations::r1cs::{ConstraintSystemRef, SynthesisError};
+use ark_std::marker::PhantomData;
+use ark_std::vec::Vec;
+
+/// A sponge that offers backwards compatibility for implementations that do not accept sponge
+/// objects but require domain separation. Operates in the same way as fork.
+#[derive(Derivative)]
+#[derivative(Clone(bound = "D: DomainSeparator"))]
+pub struct DomainSeparatedSpongeVar<
```

Comment:
I think we should avoid this struct, and just upgrade applications to use the `fork` method.

Reply:
Shall we remove it now?